core/textinput/src/textinput/Editor.cpp

```diff
@@ -90,42 +96,43 @@ namespace textinput {
     fSearch.clear();
     fUndoBuf.clear();
     CancelSpecialInputMode(R);
-    if (fReplayHistEntry != (size_t) -1) {
+    if (fReplayHistEntry != static_cast<size_t>(-1)) {
       --fReplayHistEntry; // intentional overflow to -1
       fContext->SetLine(fContext->GetHistory()->GetLine(fReplayHistEntry));
     }
     return R;
   }
 
   void
-  Editor::SetReverseHistSearchPrompt(Range& RDisplay) {
-    std::string P("[bkw'");
-    SetEditorPrompt(Text(P + fSearch + "'] "));
+  Editor::SetHistSearchModePrompt(Range& RDisplay) {
+    assert(fMode == kHistFwdSearchMode || fMode == kHistRevSearchMode);
+    const std::string direction(fMode == kHistFwdSearchMode ? "fwd" : "bkw");
+    SetEditorPrompt(Text("[" + direction + "'" + fSearch + "'] "));
     RDisplay.ExtendPromptUpdate(Range::kUpdateEditorPrompt);
   }
 
   bool
   Editor::UpdateHistSearch(EditorRange& R) {
     History* Hist = fContext->GetHistory();
     Text& Line = fContext->GetLine();
-    size_t NewHistEntry = (size_t) -1;
-    if (fSearch.empty()) {
-      NewHistEntry = 0;
-    } else {
-      size_t startAt = fCurHistEntry;
-      if (startAt == (size_t) -1) {
+    ssize_t NewHistEntry = -1;
+    if (!fSearch.empty()) {
```

Comment:
If `fSearch` is empty, the function returns `false` in line 149, which makes `Editor::ProcessCommand()` return `kPRError` (terminal bell ring).

Instead, I would suggest the following - don't forget to remove the closing brace in line 132 and adjust indentation:
```suggestion
    if (fSearch.empty())
      return true;
```